pyomo/repn/plugins/gams_writer.py

```diff
@@ -216,8 +229,18 @@ def _write_model(self,
         valid_ctypes = set([
             Block, Constraint, Expression, Objective, Param, 
             Set, RangeSet, Var, Suffix ])
+        valid_kernel_ctypes = set([
```

Comment:
The ctypes on the kernel objects match the ctypes for the AML objects. So you can just use the valid_ctypes from above, and do something like

```python
valid_ctypes = set([Block, ...])
if isinstance(model, IBlockStorage):
    if not model.collect_ctypes(active=True, descend_into=True).issubset(valid_ctypes):
        raise RuntimeError(...)
else:
    for c in model.component_objects(active=True):
        if c.type no in valid_ctypes:
            ...
```